File: 卡码/代码随想录/1/09-动态规划/Go/02-背包问题/01-01背包/Kama_0046_1.go
// 46. 携带研究材料
package main

import (
	"fmt"
	"math"
)

func main(){
	var m,n int
	fmt.Scan(&m,&n)
	weight,value:=make([]int,m),make([]int,m)
	for i:=0;i<m;i++ {
		fmt.Scan(&weight[i])
	}
	for i:=0;i<m;i++ {
		fmt.Scan(&value[i])
	}
	dp:=make([]int,n+1)
	for i:=0;i<m;i++ {
		for j:=n;j>=weight[i];j-- {
			dp[j]=int(math.Max(float64(dp[j]),float64(dp[j-weight[i]]+value[i])))
		}
	}
	fmt.Println(dp[n])
}

File: 卡码/代码随想录/1/09-动态规划/Js/02-背包问题/01-01背包/Kama_0046_1.js
//46. 携带研究材料
const rl=require("readline").createInterface({input:process.stdin});
const iter=rl[Symbol.asyncIterator]();
const readline=async ()=>(await iter.next()).value;

(async function(){
	const [m,n]=(await readline()).split(" ").map(Number);
	const weight=(await readline()).split(" ").map(Number);
	const value=(await readline()).split(" ").map(Number);
	const dp=new Array(n+1).fill(0);
	for(let i=0;i<m;i++){
		for(let j=n;j>=weight[i];j--){
			dp[j]=Math.max(dp[j],dp[j-weight[i]]+value[i]);
		}
	}
	console.log(dp[n]);
})();